### PR TITLE
Include Local monitoring tests for Datasource APIs

### PR DESCRIPTION
**Pull request by shreyabiradar07**
This PR has following changes: 

- Adds local monitoring test suite 
- Implements test cases for `/dsmetadata` API

- Test cases implemented for  Import Metadata -> POST - `/dsmetadata`
    - Importing metadata for a valid datasource to the API.
    - Post the same datasource again
    - Test with invalid values such as blank, null or an invalid value for various keys in the dsmetadata input request json
    - Validate error messages when the mandatory fields are missing

- List Datasources -> GET - `/datasources`
    - List all datasources
    - List datasources with name query parameter /datasources?name=<datasource_name>
    - List datasources with invalid parameter value for datasource name
        - tested with empty, NULL or invalid.
    - Delete a datasource and verify if list datasources doesn't list it  **(Yet to be added as delete API for datasources is not implemented)**

### CODE REVIEW

**Review by chandrams — COMMENTED** [review]
@shreyabiradar07 - Can you run the remote monitoring tests that are affected due to moving out helpers to keep it common and ensure they run fine.

**Comment by shreyabiradar07** [comment]
> @shreyabiradar07 - Can you run the remote monitoring tests that are affected due to moving out helpers to keep it common and ensure they run fine.

Test results for:
functional tests - [remote_monitoring_tests.zip](https://github.com/kruize/autotune/files/15149148/remote_monitoring_tests.zip)
fault tolerant tests - [remote-monitoring-fault-tolerant-test.zip](https://github.com/kruize/autotune/files/15149164/remote-monitoring-fault-tolerant-test.zip)

Quay image -  `quay.io/shbirada/local_monitoring:tests`

**Comment by chandrams** [comment]
@shreyabiradar07 - Remote monitoring results look fine, so common helper changes work fine. You can update the tests for the datasource API and open it for review.

**Comment by chandrams** [comment]
@shreyabiradar07 - Have you run the kruize local tests? Can you share the results

**Comment by shreyabiradar07** [comment]
> @shreyabiradar07 - Have you run the kruize local tests? Can you share the results

Test results on minikube: 
[local_monitoring_test_results.zip](https://github.com/kruize/autotune/files/15217607/local_monitoring_test_results.zip)

Jenkins job : https://ci.app-svc-perf.corp.redhat.com/job/ExternalTeams/job/Autotune/job/kruize_functional_tests/80/

**Comment by chandrams** [comment]
@shreyabiradar07 - Please update the test documentation. You also need to enable kruize local in manifests to true while running kruize local testsuite

File: tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    form_kruize_url(cluster_type)
+
+    # Get the datasources name
+    datasource_name = None
+    response = list_datasources(datasource_name)
+
+    list_datasources_json = response.json()
+
+    assert response.status_code == SUCCESS_200_STATUS_CODE
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py:26]: Please include validation of the response message as well

File: tests/README.md [hunk 2]
```diff
@@ -143,6 +143,16 @@ To run the stress test refer the Stress test [README](/tests/scripts/remote_moni
...
+### Local monitoring tests
+
+Here we test Kruize [Local monitoring APIs](/design/KruizeLocalAPI.md).
+
+#### API tests
+
+The tests does the following:
+- Deploys kruize in non-CRD mode using the deploy script from the autotune repo
+- Validates the behaviour of list datasources, import metadata and list metadata APIs in various scenarios covering both positive and negative usecases.
+
```
Comment (chandrams) [inline tests/README.md:155]: Include the link to local monitoring tests doc here similar to remote monitoring

File: tests/scripts/local_monitoring_tests/Local_monitoring_tests.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+Use the below command to test :
+
+```
+<KRUIZE_REPO>/tests/test_autotune.sh -c minikube -r [location of benchmarks]  [-i kruize image] [--tctype=functional] [--testmodule=Autotune module to be tested] [--testsuite=Group of tests that you want to perform] [--testcase=Particular test case that you want to test] [-n namespace] [--resultsdir=results directory] [--skipsetup]
+```
+
+Where values for test_autotune.sh are:
+
+```
+usage: test_autotune.sh [ -c ] : cluster type. Supported type - minikube
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/Local_monitoring_tests.md:47]: Include openshift in supported types. Specify the default option. Please update remote monitoring doc as well.

File: tests/scripts/local_monitoring_tests/rest_apis/test_import_metadata.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    assert response.status_code == expected_status_code, \
+        f"Mandatory field check failed for {field} actual - {response.status_code} expected - {expected_status_code}"
+    assert metadata_json['status'] == expected_status
+
+    response = delete_metadata(json_file)
+    print("delete metadata = ", response.status_code)
+
+
+@pytest.mark.sanity
+def test_duplicate_import_metadata(cluster_type):
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_import_metadata.py:160]: Rename the test name to something appropriate

File: tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import pytest
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py:1]: Please add copyrights wherever missing

File: tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    assert response.status_code == SUCCESS_200_STATUS_CODE
+
+    # Validate the json against the json schema
+    errorMsg = validate_list_datasources_json(list_datasources_json, list_datasources_json_schema)
+    assert errorMsg == ""
+
+
+@pytest.mark.negative
+@pytest.mark.parametrize("datasource_name", ["", "null", "xyz"])
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py:54]: Include corresponding test name as well for better readability

File: tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py (hunk 1 above)
Comment (shreyabiradar07) [reply tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py:41]: In this case accessing `list_datasources_json['message']` is resulting in  `KeyError: 'message'` as the expected response is a JSON - with list of datasources, have added validation of the output json.

File: tests/scripts/local_monitoring_tests/rest_apis/test_import_metadata.py (hunk 4 above)
Comment (shreyabiradar07) [reply tests/scripts/local_monitoring_tests/rest_apis/test_import_metadata.py]: renamed the test to ` test_repeated_metadata_import`

File: tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py (hunk 6 above)
Comment (shreyabiradar07) [reply tests/scripts/local_monitoring_tests/rest_apis/test_list_datasources.py]: done

**Review by chandrams — APPROVED** [review]
LGTM

We can add any further tests in separate PRs

**Pull request merged**